### ISSUE TITLE
Refactor application bootstrap to hide implementation detail `ReflectionCapabilities` from the user

### ISSUE THREAD

**Issue by kegluneq** [angular/angular#2304]
cc: @yjbanov , @kwalrath 

I've been talking with folks recently about the need for the user to instantiate `ReflectionCapabilities`. I remember we spoke about this a while back and that you had some concerns about hiding this from the user completely, but I can't remember the specifics and was hoping to get them documented since I get this question a lot.

The proposal is:

In application.ts
```
export function bootstrap(...) {
  if (reflector.reflectionCapabilities == null) {
    reflector.reflectionCapabilities = new ReflectionCapabilities();
  }
  bootstrapStatic(...);
}

export function bootstrapStatic(...) {
  // The current implementation of `bootstrap`
}
```

The user writes main.dart:
```
void main() {
  bootstrap(MyComponent);
}
```
which we then rewrite to:
```
import 'main.ng_deps.dart' as i0;
void main() {
  i0.initReflector();bootstrapStatic(MyComponent);
}
```

What do you think?

**Comment by yjbanov** [comment]
+1

@kegluneq do you think we could rewrite to the following instead?

```dart
import 'main.ng_deps.dart' as i0;
void main() {
  bootstrapStatic(MyComponent, i0.initReflector);
}
```

One minor issue with just calling `i0.initReflector()` is that if it fails for some reason (although it shouldn't) it would print a non-sensical stack trace pointing into code that the user didn't write. If instead we pass the `initReflector` function for `bootstrapStatic` to call, then `bootstrapStatic` could wrap the call in a `try/catch` and fail with a more meaningful message, e.g. "You are running the application with transformers enabled. The reflector generated by the transformer failed to initialize. Here is the original stack trace: ${s}".

**Comment by kegluneq** [comment]
Great point, I'm all for including this (or something similar). To avoid conflating the reflector init with bootstrap we could do:
```
wrapInitReflector(i0.initReflector);bootstrapStatic(MyComponent);
```

**Comment by kegluneq** [comment]
cc @tbosch

**Comment by kegluneq** [comment]
Discussed this offline & here's what we determined:

# TL;DR:

We can hide `ReflectionCapabilities` as an implementation detail.

# Former Situation

Previously, bindings provided to the `Injector` were immediately resolved. This meant that providing a binding to the `Injector` depended on `ReflectionCapabilities` having been instantiated. Since we needed to support providing bindings to the injector prior to calling `bootstrap`, we couldn't tie `ReflectionCapabilities` instantiation to `bootstrap`

# Current Situation

Resolution of bindings is now deferred until later in the process, so we no longer need `ReflectionCapabilities` until the user calls `bootstrap`, meaning we can hide `ReflectionCapabilities` as an implementation detail.

# Plan going forward

`bootstrap` will now automatically create `ReflectionCapabilities`. This requires `import`ing `reflection_capabilities.dart` and therefore transitively `import`ing `dart:mirrors` in the file that defines `bootstrap`. The transformer will need to remove this `import` to avoid mirrors.

## Option 1

Provide a special import for Dart `bootstrap`ing that creates `ReflectionCapabilities`, then calls into the standard `bootstrap`. The transformer rewrites the special import to the standard import, removing our transitive import of `dart:mirrors`.

### Pros

This allows us to rewrite user code instead of Angular library code, which is slightly simpler from a transformer standpoint.

### Cons

This leads to a situation where the user has to remember to import the special version of `bootstrap` rather than the standard one. If the user forgets, development will fail due to lack of reflective access and the transformer will fail to insert calls to its generated code.

## Option 2

Rewrite Angular code directly, removing the `import` of `reflection_capabilities.dart` from the library that provides `bootstrap` and adding the call to the transformer's generated code.

### Pros

Use of `bootstrap` looks the same in Js & Dart code.

### Cons

Slightly more difficult transformer implementation.


Given these pros & cons, I'd prefer to go for Option 2. I'll speak with @jakemac53 about how they do this in polymer and see if it can inform our implementation.

**Comment by kegluneq** [comment]
Just discussed with @jakemac53 and we actually like option 3 better

## Option 3

Rewrite `reflection_capabilities.dart` to call our generated code. This allows us full freedom to wrap in `try...catch` and provide friendly error messages as suggested by @yjbanov and keeps complexity low since `reflection_capabilites.dart` is being completely replaced instead of modified.

### Pros

- Use of `bootstrap` looks the same in Js & Dart code.
- Simple implementation

### Cons

None identified yet, we'll see what shakes out when we start implementing...

**Comment by yjbanov** [comment]
Discussed option 4 offline that has the pros of option 1 without the cons.

**tl;dr** move `application.ts` to `application_common.ts`, while making `application.ts` a facade; this results in identical bootstrap code between JS and Dart, and it needs minimal code rewrites by the transformer.

## Option 4

Reorg the file layout from the current:

```
angular2.ts
  export core.ts
  export EVERYTHING_ELSE

core.ts
  export src/core/application.ts

src/core/application.ts
  export function bootstrap();
  export class ApplicationRef;
```

To the following:

```
angular2.ts  // has mirrors
  export angular2_exports.ts
  export src/core/application.ts  // has mirrors

angular2_static.dart // no mirrors
  export angular2_exports.dart
  export src/core/application_common.ts  // no mirrors
  bootstrapStatic(rootComp, reflectorInitializerFn) {
    try { reflectorInitializerFn(); } catch(e, s) {  // print nice message }
    commonBootstrap(...);
  }

angular2_exports.ts  // no mirrors
  export EVERYTHING_ELSE

src/core/application.dart  // has mirrors
  bootstrap() {
    reflector = new ReflectionCapabilities();
    commonBootstrap();
  }

src/core/application.ts
  export bootstrap = commonBootstrap

// this file is the existing application.ts renamed
src/core/application_common.ts  // no mirrors
  export function bootstrap();
  export class ApplicationRef;
```

### Transformer rewrite

We rewrite this user code:

```dart
import "angular2.dart";  // has mirrors

main() {
  bootstrap(MyComp);  // instantiates ReflectorCapabilities
}
```

To this:

```dart
import "angular2_static.dart";  // no mirrors
import "ng_deps.dart" as i0;  // generated file

main() {
  bootstrapStatic(MyComp, i0.initReflector);  // no mirrors
}
```

**Comment by mhevery** [comment]
I am changing this from discussion, to something we should do. updated title/labels to reflect.

**Comment by jakemac53** [comment]
I will take this on, and bumped the priority as it is a blocker for https://github.com/angular/angular/issues/2919

**Comment by jakemac53** [comment]
One issue I realized when writing the tests for this using option #4, we would have to rewrite ALL imports of angular2.dart to angular2_static.dart, not just the import in the entry point.

Because of this I am going to instead rewrite the import to application.dart in angular2.dart to application_{static|debug|verbose}.dart (which will all in turn import application_common.dart). This is essentially the same approach but is a bit easier to manage (just one rewrite).

**Comment by jakemac53** [comment]
Although now I realize that the mirrors config only lives in your package... :(. Gonna come back to this one I think.

**Comment by jakemac53** [comment]
removed milestone and unassigned since I won't be able to get to this before I go on vacation

**Comment by yjbanov** [comment]
@jakemac53 I can take a crack at it

**Comment by yjbanov** [comment]
Status update: I'm going to try a variation of Option 4. We will have 2 files for people to import: `angular2.dart` and `bootstrap.dart`. They are equivalent, except `angular2.dart` has no `bootstrap` and does not set up reflection capabilities (and therefore does not pull in `dart:mirrors`). This way we do not need to rewrite all `angular2.dart` imports, just `bootstrap.dart` imports.

**Comment by yjbanov** [comment]
Fixed in #3078
